lib/LaTeXML/Core/Gullet.pm

```diff
@@ -271,13 +275,26 @@ sub readXToken {
         # Note: special-purpose lookup in State, for efficiency
     elsif (defined($defn = LaTeXML::Core::State::lookupExpandable($STATE, $token, $toplevel))) {
       local $LaTeXML::CURRENT_TOKEN = $token;
-      if (my $r = $defn->invoke($self)) {
-        unshift(@{ $$self{pushback} },
-          map { (!defined $_ ? ()
-              : (($r = ref $_) eq 'LaTeXML::Core::Token' ? $_
-                : ($r eq 'LaTeXML::Core::Tokens' ? @$_
-                  : Fatal('misdefined', $r, undef, "Expected a Token, got " . Stringify($_))))) }
-            @{$r}); } }
+      my $invoked   = $defn->invoke($self) || [];
+      my @expansion = ();
+      for my $exp_t (@$invoked) {
+        my $r = ref $exp_t;
+        if ($r eq 'LaTeXML::Core::Token') {
+          push @expansion, $exp_t; }
+        elsif ($r eq 'LaTeXML::Core::Tokens') {
+          push @expansion, @$exp_t; }
+        else {
+          Fatal('misdefined', $r, undef, "Expected a Token, got " . Stringify($_)); } }
+      next unless @expansion;
+      if ($$LaTeXML::Core::Token::X_THE{ $$token[0] }) {
+        # magic THE_TOKS handling, add to pushback with a single-use noexpand flag only valid
+        #    at the exact time
+        # the token leaves the pushback.
+        # This is *required to be different* from the noexpand flag, as per the B Book
+        # Let's choose something that will immediately break latexml if we overlook it...
+        @expansion = map { ['the_token', $_] } @expansion; }
```

Comment:
This comment and line of code are the crux of the "solution", as ugly as it feels. That's why I am calling it "faithful to the B Book", maximizing positive attitude.

TeX prescribes a magical global array where tokens from expansions of `\the-like primitives` are smuggled to, and then reinserted a level up at `scan_toks`. Since we don't have scan_toks, a naive approach on my part had to require rewriting all loops with `readXToken`, at the least. As it turned out, it required even more - and ultimately felt unfeasible, as the refactor was going to be huge, and possibly very fragile.

So instead of all that, I had an Eureka moment while cargo-culting boilerplate and refactoring, that this can be handled fully internally, as we initially tried with the "no_expand flag" in `$$token[2]`. But the key reason we failed is that we added the flag to the token and hoped both cases are treated similarly. They are not. I saw expl3 expansion misbehave for this reason.

The `\the`-related flag is one connected to the very first time the expanded tokens are being "consumed" from the global variable holding them. And since we always keep such tokens (coming from an invocation) into the Gullet's `{pushback}`, the "aha!" moment was that what we really want is some way to add a flag "for the duration of presence in the pushback". Crucially, that flag must always and unconditionally expire the exact moment that the token is removed from the gullet. Luckily, at least for this part, virtually all code paths end up in `readToken` and `readXToken` for removing the token from the pushback, so I can add two lines in those two subroutines and have support in the entire codebase. Whew.

The annoying bit is that I don't have a great way to avoid paying _some_ performance penalty on every pop from the pushback, to check for this case. For the first stab I used a "magical array", which is impossible to use as a Token, and has to be handled immediately (or Perl will die violently, which is good in this case - it alerts where to add more support).